pkg/cloud/rgraph/rnode/action_update.go

```diff
@@ -18,34 +18,95 @@ package rnode
 
 import (
 	"context"
+	"fmt"
+	"time"
 
 	"github.com/GoogleCloudPlatform/k8s-cloud-provider/pkg/cloud"
+	"github.com/GoogleCloudPlatform/k8s-cloud-provider/pkg/cloud/api"
 	"github.com/GoogleCloudPlatform/k8s-cloud-provider/pkg/cloud/rgraph/exec"
 )
 
-// TODO
+func UpdateActions[GA any, Alpha any, Beta any](
+	ops GenericOps[GA, Alpha, Beta],
+	got, want Node,
+	resource api.Resource[GA, Alpha, Beta],
+) ([]exec.Action, error) {
+	events, err := updatePreconditions(got, want)
+	if err != nil {
+		return nil, err
+	}
+	return []exec.Action{
+		newGenericUpdateAction(events, ops, want.ID(), resource),
+	}, nil
+}
+
+func newGenericUpdateAction[GA any, Alpha any, Beta any](
+	want exec.EventList,
+	ops GenericOps[GA, Alpha, Beta],
+	id *cloud.ResourceID,
+	resource api.Resource[GA, Alpha, Beta],
+) *genericUpdateAction[GA, Alpha, Beta] {
+	return &genericUpdateAction[GA, Alpha, Beta]{
+		ActionBase: exec.ActionBase{Want: want},
+		ops:        ops,
+		id:         id,
+		resource:   resource,
+	}
+}
+
 type genericUpdateAction[GA any, Alpha any, Beta any] struct {
 	exec.ActionBase
+	ops      GenericOps[GA, Alpha, Beta]
+	id       *cloud.ResourceID
+	resource api.Resource[GA, Alpha, Beta]
+
+	start, end time.Time
 }
 
 func (a *genericUpdateAction[GA, Alpha, Beta]) Run(
 	ctx context.Context,
 	c cloud.Cloud,
 ) (exec.EventList, error) {
-	return nil, nil
+	a.start = time.Now()
+	err := a.ops.UpdateFuncs(c).Do(ctx, "", a.id, a.resource)
+	a.end = time.Now()
+
+	// No events are emitted since the resource existed before the Action.
+	return nil, err
 }
 
 func (a *genericUpdateAction[GA, Alpha, Beta]) DryRun() exec.EventList {
+	// Update Action does not need to notify others on complete.
 	return nil
 }
 
 func (a *genericUpdateAction[GA, Alpha, Beta]) String() string {
-	return "GenericUpdateAction TODO"
+	return fmt.Sprintf("GenericUpdateAction(%v)", a.id)
+}
+
+func (a *genericUpdateAction[GA, Alpha, Beta]) Metadata() *exec.ActionMetadata {
+	return &exec.ActionMetadata{
+		Name:    fmt.Sprintf("GenericUpdateAction(%s)", a.id),
+		Type:    exec.ActionTypeUpdate,
+		Summary: fmt.Sprintf("Update %s", a.id),
+	}
 }
 
-func updatePreconditions(got, want Node) exec.EventList {
+func updatePreconditions(got, want Node) (exec.EventList, error) {
 	// Update can only occur if the resource Exists TODO: is there a case where
 	// the ambient signal for existance from Update op collides with a
 	// reference to it?
-	return nil // TODO: finish me
+	if got.State() != NodeExists && want.State() != NodeExists {
+		return nil, fmt.Errorf("node for update does not exist")
+	}
+	if want.Ownership() != OwnershipManaged {
+		return nil, fmt.Errorf("node for update is not managed")
+	}
```

Comment:
So no-ops are treated as an error?

Reply:
I'm not sure we want to check here for Ownership, as this happens in a layer above.